plugin/trino-ranger/src/main/java/io/trino/plugin/ranger/RangerSystemAccessControl.java

```diff
@@ -0,0 +1,1391 @@
+    }
+
+    private String getClientAddress(SystemSecurityContext context)
+    {
+        return context != null ? getClientAddress(context.getQueryId()) : null;
+    }
+
+    private String getClientType(SystemSecurityContext context)
+    {
+        return context != null ? getClientType(context.getQueryId()) : null;
+    }
+
+    private String getQueryText(SystemSecurityContext context)
+    {
+        return context != null ? getQueryText(context.getQueryId()) : null;
+    }
+
+    private Instant getQueryTime(SystemSecurityContext context)
+    {
+        return context != null ? getQueryTime(context.getQueryId()) : null;
+    }
+
+    private boolean hasPermission(RangerTrinoResource resource, SystemSecurityContext context, TrinoAccessType accessType, String action)
+    {
+        RangerAccessResult result = rangerPlugin.isAccessAllowed(createAccessRequest(resource, context, accessType, action));
+
+        return result != null && result.getIsAllowed();
+    }
+
+    private boolean hasPermissionForFilter(RangerTrinoResource resource, SystemSecurityContext context, TrinoAccessType accessType, String action)
+    {
+        RangerTrinoAccessRequest request = createAccessRequest(resource, context, accessType, action);
+
+        request.setResourceMatchingScope(RangerAccessRequest.ResourceMatchingScope.SELF_OR_DESCENDANTS);
+
+        RangerAccessResult result = rangerPlugin.isAccessAllowed(request, null);
+
+        return result != null && result.getIsAllowed();
+    }
+
+    private boolean hasPermission(Collection<RangerTrinoResource> resources, SystemSecurityContext context, TrinoAccessType accessType, String action)
+    {
+        boolean ret = true;
+
+        for (RangerTrinoResource resource : resources) {
+            RangerAccessResult result = rangerPlugin.isAccessAllowed(createAccessRequest(resource, context, accessType, action));
+
+            ret = result != null && result.getIsAllowed();
+
+            if (!ret) {
+                break;
+            }
+        }
+
+        return ret;
+    }
+
+    private boolean hasPermission(RangerTrinoResource resource, Identity identity, QueryId queryId, TrinoAccessType accessType, String action)
+    {
+        RangerAccessResult result = rangerPlugin.isAccessAllowed(createAccessRequest(resource, identity, queryId, accessType, action));
+
+        return result != null && result.getIsAllowed();
+    }
+
+    private boolean hasPermissionForFilter(RangerTrinoResource resource, Identity identity, QueryId queryId, TrinoAccessType accessType, String action)
+    {
+        RangerTrinoAccessRequest request = createAccessRequest(resource, identity, queryId, accessType, action);
+
+        request.setResourceMatchingScope(RangerAccessRequest.ResourceMatchingScope.SELF_OR_DESCENDANTS);
+
+        RangerAccessResult result = rangerPlugin.isAccessAllowed(request, null);
+
+        return result != null && result.getIsAllowed();
```

Comment:
why plugin can return `null`?

Reply:
This call can return null if the plugin could not get policies from Ranger and no previously downloaded policies exist in local cache directory.